language/testing-infra/transactional-test-runner/src/tasks.rs

```diff
@@ -0,0 +1,319 @@
+                    None => vec!["task", "--help"],
+                    Some(c) => vec!["task", c, "--help"],
+                };
+                let help = match Command::from_iter_safe(help_command) {
+                    Ok(_) => panic!(),
+                    Err(e) => e,
+                };
+                bail!(
+                    "Invalid command. Lines {} - {}.\n{}",
+                    start_line,
+                    command_lines_stop,
+                    help
+                )
+            }
+        };
+        let name = name_opt.unwrap();
+
+        let stop_line = if text.is_empty() {
+            command_lines_stop
+        } else {
+            text[text.len() - 1].0
+        };
+
+        // Keep fucking this up somehow
+        // let last_non_whitespace = text
+        //     .iter()
+        //     .rposition(|(_, l)| !WHITESPACE.is_match(l))
+        //     .unwrap_or(0);
+        // let initial_text = text
+        //     .into_iter()
+        //     .take_while(|(i, _)| *i < last_non_whitespace)
+        //     .map(|(_, l)| l);
+        let file_text_vec = (0..command_lines_stop)
+            .map(|_| String::new())
+            .chain(text.into_iter().map(|(_ln, l)| l))
+            .collect::<Vec<String>>();
+        let data = if file_text_vec.iter().all(|s| WHITESPACE.is_match(s)) {
+            None
+        } else {
+            let data = NamedTempFile::new()?;
+            data.reopen()?
+                .write_all(file_text_vec.join("\n").as_bytes())?;
+            Some(data)
+        };
+
+        tasks.push(TaskInput {
+            command,
+            name,
+            number,
+            start_line,
+            command_lines_stop,
+            stop_line,
+            data,
+        })
+    }
+    Ok(tasks)
+}
+
+impl<T: Debug + StructOpt> TaskInput<T> {
+    pub fn split<Subcommand: Debug + StructOpt>(
+        self,
+        split: impl FnOnce(T) -> Either<ProvidedCommand, Subcommand>,
+    ) -> Either<TaskInput<ProvidedCommand>, TaskInput<Subcommand>> {
+        let Self {
+            command,
+            name,
+            number,
+            start_line,
+            command_lines_stop,
+            stop_line,
+            data,
+        } = self;
+        match split(command) {
+            Either::Left(cmd) => Either::Left(TaskInput {
+                command: cmd,
+                name,
+                number,
+                start_line,
+                command_lines_stop,
+                stop_line,
+                data,
+            }),
+            Either::Right(subcmd) => Either::Right(TaskInput {
+                command: subcmd,
+                name,
+                number,
+                start_line,
+                command_lines_stop,
+                stop_line,
+                data,
+            }),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub enum SyntaxChoice {
+    Move,
+    IR,
+}
+
+#[derive(Debug, StructOpt)]
+pub enum ProvidedCommand {
```

Comment:
I've been trying implement a DiemTestAdapter in the past few days, and based on that experience, we might eventually need the ability to specify additional options for the provided commands. 

This is to set certain parameters for Diem transactions. Here's some code copied from `functional-tests` to give you an idea what these parameters are.
```

struct TransactionParameters<'a> {
    pub sender_addr: AccountAddress,
    pub secondary_signer_addresses: Vec<AccountAddress>,
    pub pubkey: &'a Ed25519PublicKey,
    pub privkey: &'a Ed25519PrivateKey,
    pub secondary_pubkeys: Vec<Ed25519PublicKey>,
    pub secondary_privkeys: Vec<&'a Ed25519PrivateKey>,
    pub sequence_number: u64,
    pub max_gas_amount: u64,
    pub gas_unit_price: u64,
    pub gas_currency_code: String,
    pub expiration_timestamp_secs: u64,
}
```

Reply:
I think we should somehow look at being able to provide additional commands to these in the trait. We don't want to pollute these for the VM side of things.

Reply:
> I think we should somehow look at being able to provide additional commands to these in the trait.

Hmm... do you mean adding a command to set these parameters, which will be applied to all (or maybe next) subsequent transactions? I think that could work! Although depending on the actual design, it might get a little bit messy in certain cases, when you only want these to affect 1 transaction.

> We don't want to pollute these for the VM side of things.

In general I agree, and I certainly don't think we should just add these parameters to the Run command. What I was thinking about was adding an optional string field that will be further parsed by the adapter. IMO the reason why the provided commands even exists is because they are common operations across different platforms, but I think we do have to recognize that many of these platforms will want some level of customizability.

Reply:
I'm saying you can do something of like `ProvidedCommand<ExtraArgs>` and then populate that in the macro, and tie together with an associated type in the trait. Won't be pretty but... it might work.... maybe

Or you just gut it and share less code, that works too and shouldn't be that big of deal

Reply:
> I'm saying you can do something of like ProvidedCommand<ExtraArgs> and then populate that in the macro, and tie together with an associated type in the trait. Won't be pretty but... it might work.... maybe

Good to hear that you are okay with this... at least to some extent...😆

Anyways, this adds another option we could consider. I likely won't (and shouldn't) include this in my first PR so let's not worry too much about it right now. It might become clearer to us which solution would work at that time...